core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -568,15 +568,18 @@ UNWIND(	.cantunwind)
 	cpu_is_ready
 
 #if defined (CFG_BOOT_SECONDARY_REQUEST)
-	/* generic_boot_core_hpen return value (r0) is ns entry point */
+	/* generic_boot_core_hpen return value (r0) is ns entry context structure */
 	bl	generic_boot_core_hpen
+	ldr	r6, [r0, #4]	/* load nonsecure r0 value to r6 */
+	ldr	r0, [r0, #0]	/* load entry point address to r0 */
 #else
 	mov	r0, r5		/* ns-entry address */
+	mov	r6, #0
 #endif
 	bl	generic_boot_init_secondary
 
 	mov	r0, #TEESMC_OPTEED_RETURN_ENTRY_DONE
```

Comment:
If the desired values for r0-r3 was written in `struct sm_nsec_ctx` for this cpu instead we could replace these lines with:
```
bl    sm_get_nsec_ctx
add   r0, r0, #SM_NSEC_CTX_R0
ldm   r0, {r1-r4}
mov   r0, #TEESMC_OPTEED_RETURN_ENTRY_DONE
smc   #0
```
This would be more flexible.

Reply:
It just occurred to me that this will only work when we have a secure monitor, this requires more thought.